crypto/evp/pkey_kdf.c

```diff
@@ -38,97 +42,121 @@ static int pkey_kdf_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
-        if (ret < 1)
-            return ret;
+    /* TODO: this would require a function to do properly, a side effect of
+     * a param being present is dodgy.  Instead the TLS secret param processing
+     * can reset the seed.
+        *p++ = OSSL_PARAM_construct_int(OSSL_KDF_PARAM_TLS_RESET_SEED, &one);
+     */
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_SECRET;
         break;
     case EVP_PKEY_CTRL_TLS_SEED:
-        cmd = EVP_KDF_CTRL_ADD_TLS_SEED;
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_SEED;
         break;
     case EVP_PKEY_CTRL_HKDF_KEY:
-        cmd = EVP_KDF_CTRL_SET_KEY;
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_KEY;
         break;
     case EVP_PKEY_CTRL_HKDF_INFO:
-        cmd = EVP_KDF_CTRL_ADD_HKDF_INFO;
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_INFO;
         break;
     case EVP_PKEY_CTRL_HKDF_MODE:
-        cmd = EVP_KDF_CTRL_SET_HKDF_MODE;
+        cmd = T_INT;
+        name = OSSL_KDF_PARAM_MODE_INT;
         break;
     case EVP_PKEY_CTRL_SCRYPT_N:
-        cmd = EVP_KDF_CTRL_SET_SCRYPT_N;
+        cmd = T_UINT64;
+        name = OSSL_KDF_PARAM_N;
         break;
     case EVP_PKEY_CTRL_SCRYPT_R:
-        cmd = EVP_KDF_CTRL_SET_SCRYPT_R;
+        cmd = T_UINT32;
+        name = OSSL_KDF_PARAM_R;
         break;
     case EVP_PKEY_CTRL_SCRYPT_P:
-        cmd = EVP_KDF_CTRL_SET_SCRYPT_P;
+        cmd = T_UINT32;
+        name = OSSL_KDF_PARAM_P;
         break;
     case EVP_PKEY_CTRL_SCRYPT_MAXMEM_BYTES:
-        cmd = EVP_KDF_CTRL_SET_MAXMEM_BYTES;
+        cmd = T_UINT64;
+        name = OSSL_KDF_PARAM_MAXMEM;
         break;
     default:
         return -2;
     }
 
     switch (cmd) {
-    case EVP_KDF_CTRL_SET_PASS:
-    case EVP_KDF_CTRL_SET_SALT:
-    case EVP_KDF_CTRL_SET_KEY:
-    case EVP_KDF_CTRL_SET_TLS_SECRET:
-    case EVP_KDF_CTRL_ADD_TLS_SEED:
-    case EVP_KDF_CTRL_ADD_HKDF_INFO:
-        return EVP_KDF_ctrl(kctx, cmd, (const unsigned char *)p2, (size_t)p1);
-
-    case EVP_KDF_CTRL_SET_MD:
-        return EVP_KDF_ctrl(kctx, cmd, (const EVP_MD *)p2);
-
-    case EVP_KDF_CTRL_SET_HKDF_MODE:
-        return EVP_KDF_ctrl(kctx, cmd, (int)p1);
-
-    case EVP_KDF_CTRL_SET_SCRYPT_R:
-    case EVP_KDF_CTRL_SET_SCRYPT_P:
+    case T_OCTET_STRING:
+        *p++ = OSSL_PARAM_construct_octet_string(name, (unsigned char *)p2,
+                                                 (size_t)p1);
+        break;
+
+    case T_DIGEST:
+        mdname = EVP_MD_name((const EVP_MD *)p2);
+        *p++ = OSSL_PARAM_construct_utf8_string(name, (char *)mdname,
+                                                strlen(mdname) + 1);
+        break;
+
+    case T_INT:
+        *p++ = OSSL_PARAM_construct_int(name, &p1);
+        break;
+
+    case T_UINT32:
+        /*
+         * TODO 3.0 This range check could be removed by testing at the
+         * receiving end using OSSL_PARAM_get_uint32().
+         */
```

Comment:
Make it so

Reply:
This simplifies the code quite a bit.  Pretend that it is a 64 bit integer and pass it across.
The receiving side is already grabbing the value as 32 bit and `OSSL_PARAM_get_uint32` does the all of range checking.